cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/timeseries/FactTable.java

```diff
@@ -183,138 +187,183 @@ public void delete(FactScan scan) throws Exception {
+                                      long startTs, long endTs) throws Exception {
+
+    List<TagValue> allTags = Lists.newArrayList();
+    List<TagValue> filledTags = Lists.newArrayList();
+    List<Integer> tagToFillIndexes = Lists.newArrayList();
+    for (int i = 0; i < allTagNames.size(); i++) {
+      String tagName = allTagNames.get(i);
+      if (!tagSlice.containsKey(tagName)) {
+        tagToFillIndexes.add(i);
+        allTags.add(new TagValue(tagName, null));
+      } else {
+        TagValue tagValue = new TagValue(tagName, tagSlice.get(tagName));
+        filledTags.add(tagValue);
+        allTags.add(tagValue);
+      }
+    }
+
+    byte[] startRow = codec.createStartRowKey(allTags, null, startTs, false);
+    byte[] endRow = codec.createEndRowKey(allTags, null, endTs, false);
     endRow = Bytes.stopKeyForPrefix(endRow);
-    FuzzyRowFilter fuzzyRowFilter = createFuzzyRowFilter(new FactScan(startTs, endTs, null, tags),
-                                                         startRow, true);
-    Row rowResult;
-    SortedSet<String> measureNames = Sets.newTreeSet();
+    FuzzyRowFilter fuzzyRowFilter = createFuzzyRowFilter(new FactScan(startTs, endTs, null, allTags), startRow, true);
+
+    Set<String> measureNames = Sets.newHashSet();
+    int scannedRecords = 0;
+    // todo: make configurable
     Scanner scanner = timeSeriesTable.scan(startRow, endRow, null, fuzzyRowFilter);
     try {
+      Row rowResult;
       while ((rowResult = scanner.next()) != null) {
+        scannedRecords++;
+        if (scannedRecords > MAX_RECORDS_TO_SCAN_DURING_SEARCH) {
+          break;
+        }
         byte[] rowKey = rowResult.getRow();
         // since the timestamp filter is fuzzy and also the aggregate_key is at the begging of rowKey,
         // we could match rows where (ts < startTs or ts > endTs) and the tags at beginning match,
         // we will skip those rows
-        if ((codec.getTimestamp(rowKey, codec.createColumn(startTs)) < startTs) ||
-          (codec.getTimestamp(rowKey, codec.createColumn(endTs)) > endTs)) {
+        if (codec.getTimestamp(rowKey, codec.createColumn(startTs)) < startTs) {
           continue;
         }
-        List<TagValue> tagValues = codec.getTagValues(rowResult.getRow());
-        // if the timestamp is within query range and the tag-list matches, we add them to the result
-        if (startsWithTags(tags, tagValues)) {
-          //add item to result set and increment the start rowKey
-          measureNames.add(codec.getMeasureName(rowResult.getRow()));
-          startRow = codec.getNextRowKey(rowResult.getRow(), tagValues.size());
-          scanner = timeSeriesTable.scan(startRow, endRow, null, fuzzyRowFilter);
+        if (codec.getTimestamp(rowKey, codec.createColumn(endTs)) > endTs) {
+          // we're done with scanner
+          break;
         }
+        measureNames.add(codec.getMeasureName(rowResult.getRow()));
```

Comment:
we don't update the scanner here ? tagToFillIndexes, etc are unused in this method.